packages/access/migrations/sqls/20181210164010-improvements-up.sql

```diff
@@ -0,0 +1,62 @@
+-- Function Update: Handle camel-cased table names e.g. "accessGrants".
+--
+-- First argument _tbl accepts of type regclass and requires a table class
+-- either passed with single quotes ('payments'), or double quotes if table name
+-- is camel cased ("accessGrants"). SELECT query in functon later wrapped
+-- _tbl again in double quotes which led to an error.
+--
+CREATE OR REPLACE FUNCTION make_hrid(IN _tbl regclass, IN _column text, IN digits bigint) RETURNS text AS $$
+DECLARE
+chars char[];
+new_hrid text;
+done bool;
+BEGIN
+  chars := ARRAY['1','2','3','4','5','6','7','8','9','A','B','C','D','E','F','G','H','K','L','M','R','S','T','U','W','X','Y','Z'];
+  done := false;
+  <<doneloop>>
+  WHILE NOT done LOOP
+    new_hrid := '';
+    <<hridloop>>
+    WHILE char_length(new_hrid) < digits LOOP
+      new_hrid := new_hrid || chars[floor(random()*(array_length(chars, 1)-1+1))+1];
+    END LOOP hridloop;
+    EXECUTE format('SELECT (NOT EXISTS (SELECT 1 FROM %s WHERE "%s" = %L))::bool', _tbl, _column, new_hrid) INTO done;
+  END LOOP doneloop;
+  RETURN new_hrid;
+END;
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+ALTER TABLE "accessGrants"
+  ALTER COLUMN "beginAt" DROP NOT NULL,
+  ALTER COLUMN "endAt" DROP NOT NULL,
+  ADD COLUMN "beginBefore" timestamp with time zone,
+  ADD COLUMN "voucherCode" text UNIQUE,
```

Comment:
😬

Reply:
It appears there [no gain in performance using `character varying` over `text`](https://www.depesz.com/2010/03/02/charx-vs-varcharx-vs-varchar-vs-text/), unless limited to n characters. It may have it's [roots in the past](https://serverfault.com/questions/240813/whats-difference-between-varchar-and-text-type-in-postgresql/240907#240907). In case you declare a length `character varying(5)`, Postgres will simply cut string and sture it as it would any other text type:

> All those data types are internally saved using the same C data structure – varlena.

[Limit may come at a cost](https://www.postgresql.org/docs/10/datatype-character.html):

> While character(n) has performance advantages in some other database systems, there is no such advantage in PostgreSQL; in fact character(n) is usually the slowest of the three because of its additional storage costs. In most situations text or character varying should be used instead.

Some writers mention also migrations might get much harder if altering limit, at least lowering it.

Hence, I change my stance and refuse to advocate `character varying(n)`, `varchar(n)`, `char(n)` anymore unless you have to make mission-critically-sure that there is no more than _n_ chars in a said field. Say, you fly to Mars and want to redeem a voucher code.